### PR TITLE
`unordered_set` serialization fix + tests.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: TypeSystem/Serialization/test.cc [hunk 1]
```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
...
+  EXPECT_EQ("false", JSON(false));
+  EXPECT_FALSE(ParseJSON<bool>("false"));
+  ASSERT_THROW(ParseJSON<bool>("0"), JSONSchemaException);
+  ASSERT_THROW(ParseJSON<bool>(""), InvalidJSONException);
+
+  // `int`.
+  EXPECT_EQ("42", JSON(42));
+  EXPECT_EQ(42, ParseJSON<int>("42"));
+
+  // `std::string`.
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: Also test `\n`?

File: TypeSystem/Serialization/test.cc [hunk 2]
```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
...
+  // `std::vector<>`.
+  EXPECT_EQ("[]", JSON(std::vector<uint64_t>()));
+  EXPECT_EQ("[1,2,3]", JSON(std::vector<uint64_t>({1, 2, 3})));
+  EXPECT_EQ("[[\"one\",\"two\"],[\"three\",\"four\"]]",
+            JSON(std::vector<std::vector<std::string>>({{"one", "two"}, {"three", "four"}})));
+  EXPECT_EQ(4u, ParseJSON<std::vector<std::vector<std::string>>>("[[],[],[],[]]").size());
+  EXPECT_EQ("blah", ParseJSON<std::vector<std::vector<std::string>>>("[[],[\"\",\"blah\"],[],[]]")[1][1]);
+
+  // `std::map<>`.
+  using map_int_int = std::map<int, int>;
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: Nit: `map_int_int_t`.

File: TypeSystem/Serialization/test.cc [hunk 3]
```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
...
+  using map_string_int = std::map<std::string, int>;
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: Nit: `map_string_int_t`.

File: TypeSystem/Serialization/test.cc [hunk 4]
```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
...
+  EXPECT_EQ("[]", JSON(map_int_int()));
+  EXPECT_EQ("{}", JSON(map_string_int()));
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: Would be nice to add a comment here that string-as-key results in an object, while any other key type makes it an array of arrays of two elements.

File: TypeSystem/Serialization/test.cc [hunk 5]
```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
...
+
+  EXPECT_EQ(3u, ParseJSON<map_int_int>("[[2,4],[3,9],[4,16]]").size());
+  EXPECT_EQ(16, ParseJSON<map_int_int>("[[2,4],[3,9],[4,16]]").at(4));
+  ASSERT_THROW(ParseJSON<map_int_int>("{}"), JSONSchemaException);
+
+  try {
+    ParseJSON<map_int_int>("{}");
+    ASSERT_TRUE(false);  // LCOV_EXCL_LINE
+  } catch (const JSONSchemaException& e) {
+    EXPECT_EQ("Expected map as array, got: {}", e.What());
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: A comment won't hurt here too.

File: TypeSystem/Serialization/test.cc [hunk 6]
```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
...
+  ASSERT_THROW(ParseJSON<map_string_int>("[]"), JSONSchemaException);
+  try {
+    ParseJSON<map_string_int>("[]");
+    ASSERT_TRUE(false);  // LCOV_EXCL_LINE
+  } catch (const JSONSchemaException& e) {
+    EXPECT_EQ("Expected map as object, got: []", e.What());
+  }
+
+  // `std::set<>`.
+  using set_int = std::set<int>;
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: Same two nit comments: `*_t` for `using`-s.

File: TypeSystem/Serialization/test.cc [hunk 7]
```diff
@@ -200,7 +216,129 @@ TEST(Serialization, Binary) {
...
+    ASSERT_TRUE(Exists(parsed_with_both.b));
+    EXPECT_EQ(42, Value(parsed_with_both.i));
+    EXPECT_TRUE(Value(parsed_with_both.b));
+
+    const auto parsed_with_b = LoadFromBinary<WithOptional>(ifs);
+    ASSERT_FALSE(Exists(parsed_with_b.i));
+    ASSERT_TRUE(Exists(parsed_with_b.b));
+    EXPECT_TRUE(Value(parsed_with_b.b));
+  }
+}
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: :+1:

File: TypeSystem/Serialization/test.cc [hunk 8]
```diff
@@ -323,17 +461,47 @@ TEST(Serialization, JSON) {
     EXPECT_EQ("LLandP", parsed.m.at("spock"));
     EXPECT_EQ("MTFBWY", parsed.m.at("jedi"));
   }
+  // Serializing an `std::unordered_map<>` with simple key type, which becomes a JSON object.
+  {
+    WithTrivialUnorderedMap with_unordered_map;
+    EXPECT_EQ("{\"m\":{}}", JSON(with_unordered_map));
+    with_unordered_map.m["foo"] = "fizz";
+    with_unordered_map.m["bar"] = "buzz";
+    const std::set<std::string> valid_jsons{"{\"m\":{\"bar\":\"buzz\",\"foo\":\"fizz\"}}",
```
Comment (dkorolev) [inline TypeSystem/Serialization/test.cc]: This is excellent, but comparing two `std::ordereded_map<>`-s via their own `operator==()` could work too.

**Comment by mzhurovich** [comment]
@dkorolev Let's merge it if there are no objections?

**Comment by dkorolev** [comment]
Let's.

**Pull request merged**